src/qibo/quantum_info/superoperator_transformations.py

```diff
@@ -218,6 +220,127 @@ def to_liouville(channel, order: str = "row", backend=None):
+    """
+    Converts a square matrix A ∈ ℂ^{2ⁿ × 2ⁿ} into its Pauli-Liouville representation
+    using the Fast Walsh-Hadamard Transform (FHT), based on:
+    "Pauli Decomposition via the Fast Walsh-Hadamard Transform", Sec. 3.
+
+    Args:
+        matrix (ndarray): Input matrix of shape (2^n, 2^n).
+        normalize (bool, optional): If ``True``, uses the normalized Pauli basis.
+            Defaults to ``False``.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Backend
+            to be used in the execution. If ``None``, it uses the current backend.
+
+    Returns:
+        ndarray: Matrix of shape (4^n, 4^n) with coefficients α_{r,s}
+            such that A = ∑_{r,s} α_{r,s} ⋅ P_r ⋅ P_s†.
+    """
+    backend = _check_backend(backend)
+
+    dims = operator.shape[0]
+    nqubits = int(np.log2(dims))
+
+    q_array = backend.np.arange(dims, dtype=backend.np.int64)
+    r = backend.np.arange(dims).reshape(-1, 1)
+    s = backend.np.arange(dims).reshape(1, -1)
+    r_and_s = backend.np.bitwise_and(r, s)
+    wt = hamming_weight_array(r_and_s, backend)
+    alphas = backend.np.zeros((dims, dims))
+    alphas = backend.cast(alphas, dtype=backend.np.float64)
+    for ind_row in q_array:
+        for ind_col in q_array:
+            elem = 0
+            for q_ind in q_array:
+                q_xor_r = backend.np.bitwise_xor(q_ind, ind_row)
+                state_elem = operator[q_xor_r][q_ind]
+                elem += state_elem * ((-1) ** int(wt[q_ind][ind_col]))
+            elem *= 1j ** (-int(wt[ind_row][ind_col]))
+            if backend.np.abs(elem.imag) > 1e-14 and backend.np.abs(elem.real) < 1e-14:
+                elem = 1j * elem
+
+            alphas[ind_row][ind_col] = elem.real
+    if normalize:
+        alphas /= backend.np.sqrt(dims)
+
+    return alphas / dims
+
+
+def from_pauli_liouville_fht(coeffs, backend=None):
+    """
+    Reconstructs a matrix A ∈ ℂ^{2^n × 2^n} from its FHT-based Pauli-Liouville
+    decomposition coefficients α_{r,s}, as described in:
+    "Pauli Decomposition via the Fast Walsh-Hadamard Transform", Eq. (7).
+
+    A = ∑_{r,s} α_{r,s} ⋅ P_r ⋅ P_s† ⋅ (-i)^{wt(s ∧ r)}
+
+    Args:
+        coeffs (ndarray): Matrix of shape (4^n, 4^n) with Pauli coefficients.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Backend
+            to be used in the execution. If ``None``, it uses the current backend.
+
+    Returns:
+        ndarray: Reconstructed matrix A ∈ ℂ^{2^n × 2^n}.
+    """
+    import numpy as np
+    from numpy.linalg import matrix_power
+
+    from qibo.backends import _check_backend
+
+    backend = _check_backend(backend)
+
+    dim = coeffs.shape[0]
+    n = int(np.log2(dim))
+    d = 2**n
+    A = backend.np.zeros((d, d), dtype=complex)
+
+    def int_to_binary_array(x, bits):
+        return backend.np.array([int(b) for b in format(x, f"0{bits}b")])
+
+    def hamming_weight(a, b):
+        return backend.np.sum(a * b)
```

Comment:
Same here. If what you're looking for is Hamming distance between two bitstrings, `qibo.quantum_info.utils` already has an implementation for it.